src/control/server/storage/bdev/backend.go

```diff
@@ -306,39 +310,59 @@ func detectVMD() ([]string, error) {
 	return vmdAddrs, nil
 }
 
-// cleanHugePages removes hugepage files with pathPrefix that are owned by the
-// user with username tgtUsr.
-func cleanHugePages(pathPrefix, tgtUsr string) (int, error) {
-	pattern := pathPrefix + "*"
-	matches, err := filepath.Glob(pattern)
+// getFileOwner returns the username of user who owns the file at the given
+// path. Caller should check os.IsNotExist(errors.Cause(err)) to work out if
+// the call failed because the path doesn't exist or is a directory.
+func getFileOwner(path string) (string, error) {
+	info, err := os.Stat(path)
 	if err != nil {
-		return 0, errors.Wrapf(err, "files matching %q", pattern)
+		return "", err
+	}
+	if info.IsDir() {
+		return "", os.ErrNotExist
 	}
 
-	toRemove := make([]string, 0, len(matches))
-	for _, path := range matches {
-		username, err := common.GetFileOwner(path)
+	uid := info.Sys().(*syscall.Stat_t).Uid
+	uidStr := strconv.FormatUint(uint64(uid), 10)
+
+	usr, err := user.LookupId(uidStr)
+	if err != nil {
+		return "", err
+	}
+
+	return usr.Username, nil
+}
+
+// cleanHugePages removes hugepage files with pathPrefix that are owned by the
+// user with username tgtUsr.
+func cleanHugePages(hugePageDir, prefix, tgtUsr string) (int, error) {
+	var nrRemoved int
+	err := filepath.Walk(hugePageDir, func(path string, info os.FileInfo, err error) error {
 		if err != nil {
-			if os.IsNotExist(errors.Cause(err)) {
-				continue
-			}
-			return 0, err
+			return err
+		}
+		if info.IsDir() && path != hugePageDir {
+			return filepath.SkipDir // skip subdirectories
+		}
+		if !strings.HasPrefix(info.Name(), prefix) {
+			return nil // skip files without prefix
+		}
+		username, err := getFileOwner(path)
```

Comment:
Don't you already have access to the underlying `Stat()` details via `info.Sys()`? I was under the impression that `filepath.Walk()` does a `Stat()` for you, making some of the logic in `getFileOwner()` redundant.

I was thinking that you could make your WalkFunc implementation a named function that you could then unit test by supplying it with synthetic inputs to test the logic.

Reply:
attempted to address the comment